text/0000-output-options.md

```diff
@@ -0,0 +1,184 @@
+[dirname]: http://doc.rust-lang.org/std/path/trait.GenericPath.html#tymethod.dirname
+
+If both `--out-dir` and `-o` are provided, the `-o` value is adjusted so it only contains a
+[`filename`], which is used as `[filename]`:
+
+    $ rustc foo.rs --out-dir=baz -o foo # A-OK
+    $ rustc foo.rs --out-dir=baz -o bar/foo
+    warning: `-o foo` is used instead of `-o bar/foo`, because `--out-dir=baz` is specified
+
+Note, that this proposed behaviour is completely different than status quo. At the time of writing
+both commands in the previous example completely ignore `--out-dir` and write the files to `./foo`
+and `./bar/foo` respectively. This is, arguably, completely unintuitive and requires extensive
+documentation for the options. With the change both options’ behaviour becomes as intuitive as it
+gets: `--out-dir` is the option to specify output directory and `-o` is the option to specify
+output filename. Ability to specify output directory with `-o` is only a (sometimes unavailable)
+shorthand convenience.
+
+### Interactions with `--emit`
+
+The extension has no special meaning in `-o` option values and becomes a part of `[filename]`
+verbatim:
+
+    $ rustc foo.rs --emit=asm,obj -o foo.bar
+    # Output: foo.bar.s foo.bar.o
+    $ rustc foo.rs --crate-type=staticlib,rlib -o foo.bar
+    # Output: libfoo.bar.rlib libfoo.bar.a
+    $ rustc foo.rs --emit=asm -o foo.bar
+    # Output: foo.bar
+    $ rustc foo.rs --crate-type=rlib -o foo.bar
+    # Output: libfoo.bar.rlib
+
+### `-o` is not specified
+
+In case `-o` is not specified, compiler generates `[filename]` on a best-effort basis. rustc could
+use any of following sources of data:
+
+* `crate-name`;
+* [Filestem][filestem] of the input file;
+* …
+
+[filestem]: http://doc.rust-lang.org/std/path/trait.GenericPath.html#method.filestem
+
+If `[filename]` cannot be generated from the available data, a sensible default such as `rust_out`
+is used.
+
+## `-C extra-filename`
+
+When `-C extra-filename` option is specified, the `[filename]` is mutated so the new value is
+`[filename][extra-filename]`.
+
+    $ rustc foo.rs -C extra-filename=qux -o foo.bar --crate-type=rlib
+    # Output: libfoo.barqux.rlib
+
+# Drawbacks
+
+The special case for `link` target might be somewhat confusing:
+
+    $ rustc foo.rs --emit=link -o foo.bar
+    # Output on Windows: foo.bar.exe
+
+but
+
+    $ rustc foo.rs --emit=asm -o foo.bar
+    # Output: foo.bar
+
+It does not allow setting precise filename for `link` target outputs, since both `lib` prefix and
+an extension might be appended, while other targets don’t share such a restriction. On the other
+hand, the drawback is necessary, because rustc needs both the extension and the prefix to consider
+it as a linkage candidate and binaries without `.exe` extension look silly on Windows.
+
+# Alternatives
+
+Keeping current inconsistent behaviour.
+
+---
+
+When resolving conflicts betweeen [`--out-dir` and `-o` options](#interactions-with---out-dir) the
```

Comment:
I’m more concerned about behaviours rather than implementation… e.g. one might want it not go above the directories specified in `--out-dir` (`--outdir=~ -o ../../hello` → `[out-dir] = /`).

I, again, might just be looking for non-existent problems 😐